libcrux-ml-dsa/src/simd/portable/encoding/commitment.rs

```diff
@@ -54,37 +12,24 @@ Seq.length ${serialized_future} == Seq.length ${serialized} /\
 ")
 })]
 fn serialize_4(simd_unit: &Coefficients, serialized: &mut [u8]) {
-    // The commitment has coefficients in [0,15] => each coefficient occupies
-    // 4 bits.
-    cloop! {
-        for (i, coefficients) in simd_unit.values.chunks_exact(2).enumerate() {
-            hax_lib::loop_invariant!(|i: usize| {fstar!(r"
-            Seq.length $serialized == 4 /\ (
-              let inp = bit_vec_of_int_t_array #I32 #(mk_usize 8) ${simd_unit.values} 4 in
-              let out = bit_vec_of_int_t_array #U8 #(mk_usize 4) $serialized 8 in
-              forall (n: nat {n < v i * 8}). out n == inp n
-            )")});
+    let coefficient0 = simd_unit.values[0] as u8;
+    let coefficient1 = simd_unit.values[1] as u8;
+    let coefficient2 = simd_unit.values[2] as u8;
+    let coefficient3 = simd_unit.values[3] as u8;
+    let coefficient4 = simd_unit.values[4] as u8;
+    let coefficient5 = simd_unit.values[5] as u8;
+    let coefficient6 = simd_unit.values[6] as u8;
+    let coefficient7 = simd_unit.values[7] as u8;
 
-            serialized[i] = encode_4(coefficients);
+    let byte0 = (coefficient1 << 4) | coefficient0;
+    let byte1 = (coefficient3 << 4) | coefficient2;
+    let byte2 = (coefficient5 << 4) | coefficient4;
+    let byte3 = (coefficient7 << 4) | coefficient6;
 
-            hax_lib::fstar!(r"
-                let inp = bit_vec_of_int_t_array #I32 #(mk_usize 8) ${simd_unit.values} 4 in
-                let out = bit_vec_of_int_t_array #U8  #(mk_usize 4) $serialized 8 in
-                introduce forall (n:nat{n < 8}). inp (v i * 8 + n) == out (v i * 8 + n)
-                with
-                  (calc (==) {
-                          inp (v i * 8 + n);
-                    == {} get_bit (Seq.index ${simd_unit.values} ((v i * 8 + n) / 4)) (sz ((v i * 8 + n) % 4));
-                    == { Math.Lemmas.division_addition_lemma n 4 (v i * 2) }
-                          get_bit (Seq.index ${simd_unit.values} (v i * 2 + n / 4)) (sz (n % 4));
-                    == {} get_bit (Seq.index $coefficients (n / 4)) (sz (n % 4));
-                    == {} bit_vec_of_int_t_array #I32 #(mk_usize 2) $coefficients 4 n;
-                    == {} out (v i * 8 + n);
-                  })
-            ");
-        }
-    }
-    ()
+    serialized[0] = byte0;
+    serialized[1] = byte1;
+    serialized[2] = byte2;
+    serialized[3] = byte3;
```

Comment:
Interesting. I also think there will be run-time performance benefits when unrolling complicated loops, but the question is how convenient is unrolling such loops.

Reply:
I would like such unrollings to be done automatically in hax or elsewhere. Essentially, the effect of this unrolling is to make the proof trivial for SMT, and whichever way we can get this effect is fine, but would need some experimentation in coming PRs. Whatever works, we can backport to ML-KEM.